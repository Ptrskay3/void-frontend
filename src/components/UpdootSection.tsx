import { ChevronUpIcon, ChevronDownIcon } from "@chakra-ui/icons";
import { Flex, IconButton, Text } from "@chakra-ui/react";
import React, { useState } from "react";
import { PostSnippetFragment, useVoteMutation } from "../generated/graphql";

interface UpdootSectionProps {
  post: PostSnippetFragment | {}; // ugly, but pages/post/[id] needs the full post, not the fragment
  disableVoting?: boolean;
}

const UpdootSection: React.FC<UpdootSectionProps> = ({
  post,
  disableVoting = false,
}) => {
  const [loadingState, setLoadingState] =
    useState<"updoot-loading" | "downdoot-loading" | "not-loading">(
      "not-loading"
    );
  const [{}, vote] = useVoteMutation();
  return (
    <Flex direction="column" alignItems="center" justifyContent="center" mr="4">
      <IconButton
        onClick={async () => {
          // if (post.voteStatus === 1) {
          //   return;
          // }
          setLoadingState("updoot-loading");
          await vote({ postId: (post as PostSnippetFragment).id, value: 1 });
          setLoadingState("not-loading");
        }}
        aria-label="upvote"
        isLoading={loadingState === "updoot-loading"}
        icon={<ChevronUpIcon />}
        colorScheme={
          (post as PostSnippetFragment).voteStatus === 1 ? "green" : undefined
        }
        isDisabled={disableVoting}
      />
      <Text>{(post as PostSnippetFragment).points}</Text>
      <IconButton
        aria-label="downvote"
        onClick={async () => {
          // if (post.voteStatus === -1) {
          //   return;
          // }
          setLoadingState("downdoot-loading");
          await vote({ postId: (post as PostSnippetFragment).id, value: -1 });
          setLoadingState("not-loading");
        }}
        icon={<ChevronDownIcon />}
        isLoading={loadingState === "downdoot-loading"}
        colorScheme={
          (post as PostSnippetFragment).voteStatus === -1 ? "red" : undefined
        }
        isDisabled={disableVoting}
      />
    </Flex>
  );
};

export default UpdootSection;
