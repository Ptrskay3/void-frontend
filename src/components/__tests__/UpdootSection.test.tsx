import React from "react";
import { shallow } from "enzyme";
import UpdootSection from "../UpdootSection";
import { Provider } from "urql";
import { never } from "wonka";

describe("Layout", () => {
  it("Updoot section loads with post", () => {
    const mockClient = {
      executeQuery: jest.fn(() => never),
      executeMutation: jest.fn(() => never),
      executeSubscription: jest.fn(() => never),
    };

    const post = {
      points: 1,
      id: 345,
      title: "szia",
      createdAt: "1615645506657",
      updatedAt: "1615645506657",
      textSnippet: "",
      voteStatus: 1,
      creator: {
        username: "ben",
        id: 2,
      },
    };

    const updootSection = shallow(
      <Provider value={mockClient}>
        <UpdootSection post={post} />
      </Provider>
    );

    const foundPost = updootSection.find("post");
    expect(foundPost).toBeDefined();

    expect(updootSection.getElements()).toMatchSnapshot();
  });
});
