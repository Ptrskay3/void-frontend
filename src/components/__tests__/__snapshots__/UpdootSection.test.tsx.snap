// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Updoot section loads with post 1`] = `
Array [
  <UpdootSection
    post={
      Object {
        "createdAt": "1615645506657",
        "creator": Object {
          "id": 2,
          "username": "ben",
        },
        "id": 345,
        "points": 1,
        "textSnippet": "",
        "title": "szia",
        "updatedAt": "1615645506657",
        "voteStatus": 1,
      }
    }
  />,
]
`;
